source/common/router/router.cc

```diff
@@ -541,37 +578,85 @@ void Filter::onRequestComplete() {
   }
 }
 
-void Filter::onDestroy() {
-  if (upstream_requests_.size() == 1 && !attempting_internal_redirect_with_complete_stream_) {
-    upstream_requests_.front()->resetStream();
-  }
-  cleanup();
-}
+void Filter::onDestroy() { cleanup(); }
 
 void Filter::onResponseTimeout() {
   ENVOY_STREAM_LOG(debug, "upstream timeout", *callbacks_);
-  cluster_->stats().upstream_rq_timeout_.inc();
 
-  ASSERT(upstream_requests_.size() <= 1);
-  if (upstream_requests_.size() == 1) {
-    if (upstream_requests_.front()->upstream_host_) {
-      upstream_requests_.front()->upstream_host_->stats().rq_timeout_.inc();
-    }
+  // Reset any upstream requests that are still in flight.
+  for (auto& upstream_request : upstream_requests_) {
+    // Don't record a timeout for upstream requests we've already seen headers
+    // for.
+    if (!upstream_request->upstream_headers_) {
+      cluster_->stats().upstream_rq_timeout_.inc();
+      if (upstream_request->upstream_host_) {
+        upstream_request->upstream_host_->stats().rq_timeout_.inc();
+      }
+
+      // If this upstream request already hit a "soft" timeout, then it
+      // already recorded a timeout into outlier detection. Don't do it again.
+      if (!upstream_request->outlier_detection_timeout_recorded_) {
+        updateOutlierDetection(timeout_response_code_, *upstream_request);
+      }
+      upstream_request->resetStream();
 
-    updateOutlierDetection(timeout_response_code_, *upstream_requests_.front().get());
-    upstream_requests_.front()->resetStream();
+      chargeUpstreamAbort(timeout_response_code_, false, *upstream_request);
+    }
   }
 
   onUpstreamTimeoutAbort(StreamInfo::ResponseFlag::UpstreamRequestTimeout);
 }
 
+// Called when the per try timeout is hit but we didn't reset the request
+// (hedge_on_per_try_timeout enabled).
+void Filter::onSoftPerTryTimeout(UpstreamRequest& upstream_request) {
+  // Even though we didn't cancel the request yet we still want to track it
+  // in outlier detection.
+  // TODO(mpuncel) is it weird to have a pretend response code here? we might
```

Comment:
I think it's probably fine, given that we have seen what we think is an outlier from this host, even if it shows up late later. We don't charge it as a success later, right (per the boolean)? If you agree I would just remove this TODO.

Reply:
sounds good to me